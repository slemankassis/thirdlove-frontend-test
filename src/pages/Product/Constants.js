module.exports = {
  QTY_CHARS_FOR_BAND_SIZES: 2,
  COLOR_VARIANTS: 'option1',
  SIZE_VARIANTS: 'option2',
  STOCK: 'inventory_quantity',
  ID: 'id',
  TITLE: 'title',
  PRICE: 'price',
  THUMBNAIL: 'src100',
  ORIGINAL_MEDIUM: 'src600',
  ORIGINAL_LARGE: 'src1000',
};
