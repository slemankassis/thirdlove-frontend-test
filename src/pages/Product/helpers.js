import {
  COLOR_VARIANTS,
  SIZE_VARIANTS,
  QTY_CHARS_FOR_BAND_SIZES,
  ID,
  PRICE,
  STOCK,
  THUMBNAIL,
  ORIGINAL_LARGE,
  COLORS,
  NUMBER_OF_ATTRS_VARIANTS,
  MIN_STOCK,
} from './Constants';

const getColorFiltersFromVariants = (variants) => {
  const colorFilters = new Set();
  variants.forEach((variant) => {
    colorFilters.add(variant.color);
  });
  return [...colorFilters];
};

const transformImages = (images) => (
  images.reduce((transformedImages, image) => (
    transformedImages.concat({
      thumbnail: `https://${image[THUMBNAIL]}`,
      original: `https://${image[ORIGINAL_LARGE]}`,
    })
  ), [])
);

// TODO: Make friendly this func
const variantHasValidData = (variant) => (
  Object.values(variant).length === NUMBER_OF_ATTRS_VARIANTS
);

const variantHasStock = (variant) => (variant[STOCK] >= MIN_STOCK);

// Transform keys and remove invalid elements
const transformVariants = (variants) => (
  variants.reduce((transformedVariants, variant) => (
    (variantHasStock(variant) && variantHasValidData(variant))
      ? transformedVariants.concat({
        id: variant[ID].toString(),
        price: variant[PRICE],
        stock: variant[STOCK],
        color: variant[COLOR_VARIANTS],
        [SIZE_VARIANTS]: variant[SIZE_VARIANTS],
        band: `${variant[SIZE_VARIANTS][0]}${variant[SIZE_VARIANTS][1]}`,
        cup: variant[SIZE_VARIANTS].slice(QTY_CHARS_FOR_BAND_SIZES),
      })
      : transformedVariants
  ), [])
);

// TODO: Make generic with n filters
const getVariant = ({
  variants,
  color,
  band,
  cup,
}) => (
  variants && variants.find((variant) => (
    (!color || variant.color === color)
      && (!band || variant.band === band)
      && (!cup || variant.cup === cup)
  ))
);

const getColorsInHex = (colorsArray) => colorsArray.map((color) => COLORS[color]);

export {
  transformImages,
  transformVariants,
  getVariant,
  getColorsInHex,
  getColorFiltersFromVariants,
};
