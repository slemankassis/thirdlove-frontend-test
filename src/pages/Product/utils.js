import {
  COLOR_VARIANTS,
  SIZE_VARIANTS,
  QTY_CHARS_FOR_BAND_SIZES,
  ID,
  PRICE,
  STOCK,
  THUMBNAIL,
  ORIGINAL_LARGE,
} from './Constants';

const transformImages = (images) => (
  images.reduce((transformedImages, image) => (
    transformedImages.concat({
      thumbnail: `https://${image[THUMBNAIL]}`,
      original: `https://${image[ORIGINAL_LARGE]}`,
    })
  ), [])
);

const transformVariants = (variants) => (
  variants.reduce((transformedVariants, variant) => (variant[STOCK] >= 10
    ? transformedVariants.concat({
      id: variant[ID],
      price: variant[PRICE],
      stock: variant[STOCK],
      color: variant[COLOR_VARIANTS],
      [SIZE_VARIANTS]: variant[SIZE_VARIANTS],
      band: `${variant[SIZE_VARIANTS][0]}${variant[SIZE_VARIANTS][1]}`,
      cup: variant[SIZE_VARIANTS].slice(QTY_CHARS_FOR_BAND_SIZES),
    })
    : transformedVariants
  ), [])
);

const getSelectedVariant = (variants, selectedColor, selectedBand, selectedCup) => (
  variants.filter((variant) => (
    (variant.color === selectedColor)
    && (variant.band === selectedBand)
    && (variant.cup === selectedCup)
  ))
);

const getObjsFromArrayByKey = (array, id) => array.filter((element) => element.id === id);

export {
  transformImages,
  transformVariants,
  getSelectedVariant,
  getObjsFromArrayByKey,
};
