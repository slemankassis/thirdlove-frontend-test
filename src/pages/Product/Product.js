import React from 'react';
import Button from '../../thirdy-part-components/Button';
import Label from '../../thirdy-part-components/Label';
import Carousel from './components/Carousel';
import Variants from './components/Variants';
import Description from './components/Description';
import { transformImages, transformVariants, getObjsFromArrayByKey } from './helpers';
import './Product.scss';

class Product extends React.Component {
  constructor(props) {
    super(props);

    this.onchangeVariant = this.onchangeVariant.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.addToCart = this.addToCart.bind(this);

    this.state = {
      selectedVariantId: this.props.product.variants[0].id,
    };
  }

  onchangeVariant(value) {
    this.setState(() => ({
      selectedVariantId: value,
    }));
  }

  handleSubmit({
    title,
    productId,
    band,
    cup,
    variantId,
  }) {
    // eslint-disable-next-line no-alert
    alert(`Added a ${title} - ${band}${cup} to the cart`);
    this.addToCart(productId, variantId);
  }

  // eslint-disable-next-line no-unused-vars
  addToCart(productId, varianId) { // eslint-disable-line class-methods-use-this
    // TODO: Call service with productId and variantId and remove eslint exceptions
  }

  render() {
    const {
      product: {
        id: productId,
        title,
        images,
        variants,
        body_html: bodyHtml,
      },
    } = this.props;
    const { selectedVariantId } = this.state;

    const selectedVariant = getObjsFromArrayByKey(variants, selectedVariantId);
    if (!selectedVariant) {
      throw new Error('Unexpected error getting variant of product');
    }

    const transformedVariant = transformVariants([selectedVariant]);
    const {
      color,
      price,
      stock,
      band,
      cup,
    } = transformedVariant;

    return (
      <React.Fragment>
        <Carousel images={transformImages(images)} />
        {/* TODO: Use form with input dropdown and radios */}
        {/* <form onSubmit={() => this.handleSubmit(title, id, variants[selectedVariantId])}> */}
        <Label className="product-title" text={title} />
        {variants[selectedVariantId] && (
          <Label className="product-color" text={`COLOR: ${color}`} />
        )}
        <Label className="product-price" text={price} />
        <Variants
          selectedVariantId={selectedVariantId}
          onchangeVariant={this.onchangeVariant}
          variants={transformVariants(variants)}
        />
        <Label className="product-stock" text={`STOCK: ${stock}`} />
        {/* </form> */}
        <Button
          type="action"
          text="Add to Bag"
          onClick={() => this.handleSubmit({
            title,
            productId,
            band,
            cup,
            selectedVariantId,
          })}
        />
        <Description className="product-description" contentHtml={bodyHtml} />
      </React.Fragment>
    );
  }
}

export default Product;
