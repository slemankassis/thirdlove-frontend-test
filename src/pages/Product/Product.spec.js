import React from 'react';
import Product from './Product';

describe('Product snaps', () => {
  it('renders snapshot', () => {
    const props = {
      product: mockData.product,
    };

    const component = shallow(<Product {...props} />);
    expect(escapeSnapshot(component)).toMatchSnapshot();
  });
});
