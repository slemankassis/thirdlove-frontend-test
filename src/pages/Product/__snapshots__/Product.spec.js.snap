// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product renders snapshot 1`] = `
"<Fragment>
  <Carousel images={{...}} />
  <Label className='product-title' text='24/7™ Classic Perfect Coverage Bra' />
  <Label className='product-color' text='COLOR: naked-1' />
  <Label className='product-price' text='$68' />
  <Variants selectedVariantId='6989569458233' onChangeVariant={[Function: bound onChangeVariant]} variants={{...}} />
  <Label className='product-stock' text='STOCK: 764' />
  <Button text='Add to Bag' onClick={[Function: onClick]} className='' />
  <Description className='product-description' contentHtml='<meta charset='utf-8'>\\\\n<p>A slightly fuller coverage version of the bra that started it all. The 24/7™ Classic Full Coverage Bra’s signature cups are designed with hybrid memory foam that does double duty, providing softness inside and support outside. It forms to your unique shape and creates a smooth silhouette, no matter what you’re wearing. The straps are lined with supple memory foam so they stay put and never dig in, and the ballet back design ensures your straps will truly never slip. You’ll wonder how you ever lived without it.</p>\\\\n<!-- split -->\\\\n<ul>\\\\n<li><span>Available in B-H, 32-46 bands</span></li>\\\\n<li><span>Ultra-thin hybrid foam, has memory foam on the inside to form to your body, and a supportive foam on the outside</span></li>\\\\n<li><span>Incredibly soft micro jersey fabric, knitted for maximum durability</span></li>\\\\n<li><span>Double layer ballet back with hidden elastic smooths your silhouette</span></li>\\\\n<li><span>Memory foam straps that won\\\\'t slip </span></li>\\\\n<li>Pleated detail at the center front</li>\\\\n<li><span>Gold alloy hardware adds a touch of shine</span></li>\\\\n<li><span>Foam-padded hook &amp; eye with tagless (scratch-free!) printed label</span></li>\\\\n<li><span>Flexible nylon-coated nickel-free wires</span></li>\\\\n<li><span>Nylon/spandex</span></li>\\\\n</ul>' />
</Fragment>"
`;

exports[`Product snaps renders snapshot 1`] = `
"<Fragment>
  <Carousel images={{...}} />
  <Label className='product-title' text='24/7™ Classic Perfect Coverage Bra' />
  <Label className='product-color' text='COLOR: naked-1' />
  <Label className='product-price' text='$68' />
  <Variants selectedVariantId='6989569458233' onChangeVariant={[Function: bound onChangeVariant]} variants={{...}} />
  <Label className='product-stock' text='STOCK: 764' />
  <Button text='Add to Bag' onClick={[Function: onClick]} className='' />
  <Description className='product-description' contentHtml='<meta charset='utf-8'>\\\\n<p>A slightly fuller coverage version of the bra that started it all. The 24/7™ Classic Full Coverage Bra’s signature cups are designed with hybrid memory foam that does double duty, providing softness inside and support outside. It forms to your unique shape and creates a smooth silhouette, no matter what you’re wearing. The straps are lined with supple memory foam so they stay put and never dig in, and the ballet back design ensures your straps will truly never slip. You’ll wonder how you ever lived without it.</p>\\\\n<!-- split -->\\\\n<ul>\\\\n<li><span>Available in B-H, 32-46 bands</span></li>\\\\n<li><span>Ultra-thin hybrid foam, has memory foam on the inside to form to your body, and a supportive foam on the outside</span></li>\\\\n<li><span>Incredibly soft micro jersey fabric, knitted for maximum durability</span></li>\\\\n<li><span>Double layer ballet back with hidden elastic smooths your silhouette</span></li>\\\\n<li><span>Memory foam straps that won\\\\'t slip </span></li>\\\\n<li>Pleated detail at the center front</li>\\\\n<li><span>Gold alloy hardware adds a touch of shine</span></li>\\\\n<li><span>Foam-padded hook &amp; eye with tagless (scratch-free!) printed label</span></li>\\\\n<li><span>Flexible nylon-coated nickel-free wires</span></li>\\\\n<li><span>Nylon/spandex</span></li>\\\\n</ul>' />
</Fragment>"
`;
